### PR TITLE
Adds support for knowledge distillation

### PR DESCRIPTION
**Pull request by RobotSail**
There are many different forms of model training which exist. One popular form of training is knowledge distillation, where a student model learns the output distributions from a teacher model.  This commit introduces support for knowledge distillation in the training library.

This commit also exposes the `weight_decay` hyperparameter which is often used to help deep learning models generalize.

Lastly, this commit changes the useage from `torch.distributed` to just `dist`, as it is a common module used throughout the codebase.

Signed-off-by: Oleg S <97077423+RobotSail@users.noreply.github.com>

### CODE REVIEW

File: src/instructlab/training/config.py [hunk 1]
```diff
@@ -121,6 +121,17 @@ class DeepSpeedOptions(BaseModel):
...
 
+# public API
+class DistillationConfig(BaseModel):
+    """
+    Config to use when performing knowledge distillation during training.
+    """
+
+    temperature: float = Field(1.0, gt=0.0)
+    alpha: float = Field(1.0, le=1.0, ge=0.0)
+    teacher_path: str
```
Comment (JamesKunstle) [inline src/instructlab/training/config.py:132]: if possible would love to standardize on using pathlib.Path rather than str paths.

File: src/instructlab/training/main_ds.py [hunk 2]
```diff
@@ -252,6 +260,38 @@ def make_inputs_require_grad(module, input, output):
...
+def setup_teacher_model(
+    model_name_or_path: str, device: torch.device, tokenizer: PreTrainedTokenizer
+):
+    """
+    Instantiates a teacher model to be used for distillation training.
+    """
+    teacher_model = AutoModelForCausalLM.from_pretrained(
+        model_name_or_path, torch_dtype=torch.bfloat16
+    ).to(device)
+    model_dev = next(teacher_model.parameters()).device
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:272]: If you're calling `.to(device)` just above, could you make a note of why you also need to confirm the device locale below?

File: src/instructlab/training/config.py [hunk 3]
```diff
@@ -179,6 +190,11 @@ class TrainingArgs(BaseModel):
     is_padding_free: bool = False  # TODO: deprecate
     checkpoint_at_epoch: bool = True
     accelerate_full_state_at_epoch: bool = True
+    weight_decay: float = Field(0.0, ge=0.0)
+
+    # settings for knowledge distillation
+    distillation_options: Optional[DistillationConfig] = None
```
Comment (JamesKunstle) [inline src/instructlab/training/config.py:196]: I've seen that `Optional[DistillationConfig]` syntax is replaced by `DistillationConfig | None` in recent Pythonic parlance once the optional annotation was added to the language. This is a nit, not required to change.

File: src/instructlab/training/main_ds.py [hunk 4]
```diff
@@ -382,11 +454,36 @@ def train(
...
+            output: CausalLMOutput = model(
                 **batch,
                 use_cache=False,
             )
-            loss = output.loss
+
+            loss = None
+            if args.distill:
+                # teacher_model should always be provided when `args.distill` is enabled
+                if TYPE_CHECKING:
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:467]: I think this is supposed to be a runtime check but TYPE_CHECKING is always False at runtime.
https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Comment (JamesKunstle) [reply src/instructlab/training/main_ds.py:467]: I think we should fail much earlier if distillation is set but no teacher_model is provided, like before we do any data preprocessing or fire up the GPUs.

File: src/instructlab/training/main_ds.py [hunk 5]
```diff
@@ -382,11 +454,36 @@ def train(
...
+                    assert (
+                        teacher_model is not None
+                    ), "teacher model cannot be None when `distill` is enabled"
+
+                with torch.no_grad():
+                    teacher_output: CausalLMOutput = teacher_model(
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:473]: You turn off `requires_grad` on all the params in the teacher_model. You could just be doing this instead, I think this gives the same output.

File: src/instructlab/training/main_ds.py [hunk 6]
```diff
@@ -382,11 +454,36 @@ def train(
...
+                    student_output=output,
+                    teacher_output=teacher_output,
+                    alpha=args.distill_alpha,
+                    temp=args.distill_temp,
+                )
+
+            else:
+                loss = output.loss
+
+            assert loss is not None, "loss cannot be equal to None!"
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:486]: `assert`s are typically not preferred in comparison to runtime exceptions.
```suggestion
            if loss is None:
                raise ValueError("loss was None during distillation training. Something unrecoverable went wrong.")
```

Comment (JamesKunstle) [reply src/instructlab/training/main_ds.py:486]: mostly because they can be removed with `-0` when the interpreter is invoked. But we want to check non-null all the time.

File: src/instructlab/training/config.py (hunk 1 above)
Comment (RobotSail) [reply src/instructlab/training/config.py:132]: @JamesKunstle I see your point, would it make sense for it to be a path when it can also take on a HF reference? I understand that references can technically still be paths, but to a consumer reading it might sound like only local models are accepted. Would `str | Path` be satisfactory?

File: src/instructlab/training/main_ds.py (hunk 2 above)
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:272]: Yes of course.

File: src/instructlab/training/config.py (hunk 3 above)
Comment (RobotSail) [reply src/instructlab/training/config.py:196]: Let's do the proposed method to be more consistent with how Python expects optionals in the future.

File: src/instructlab/training/main_ds.py (hunk 6 above)
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:486]: Sure I can change this. I was using them as scaffolding when writing this.

File: src/instructlab/training/main_ds.py [hunk 7]
```diff
@@ -511,6 +609,9 @@ def main(args):
     # Third Party
     import yaml
 
+    if args.distill and not args.teacher_model_name_or_path:
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:612]: Sweet :party-cat:

File: src/instructlab/training/main_ds.py (hunk 4 above)
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:467]: Yeah it is, I believe I had errors with type-checking here though and it not knowing that `teacher_model` is properly set.

File: src/instructlab/training/main_ds.py (hunk 6 above)
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:486]: I'm gonna make it not be specific to distillation training though since it's more about how we branch out. I suspect as we add other loss calculations (contrastive loss, preference tuning loss, etc.), we will start out by setting it to `None` and having this final check to ensure it was set to something.
